### PR TITLE
Guard against decorator siblings in $shrinkSelectionPastBlockEdges

### PR DESCRIPTION
**Pull request by jeremy**
## Symptom

Every block-format toolbar command — Heading, Quote, Code block, Paragraph, List (call sites in `src/editor/contents.js`) — throws `TypeError: n.getChildrenSize is not a function` and silently aborts when the selection borders an attachment or horizontal divider. The toolbar button does nothing; no visible error for the user.

## Trigger

A selection spanning multiple blocks whose focus sits flush at the start of a block whose previous top-level sibling is a block-level `DecoratorNode` — a top-level `ActionTextAttachmentNode` / `CustomActionTextAttachmentNode` (`isInline()` is false when the parent is root) or a `HorizontalDividerNode` — then any block-format toolbar click.

## Root cause

In `$shrinkSelectionPastBlockEdges` (`src/helpers/lexical_helper.js`):

```js
const previousBlock = focusBlock.getPreviousSibling()
if (previousBlock) selection.focus.set(previousBlock.getKey(), previousBlock.getChildrenSize(), "element")
```

`getChildrenSize()` is an `ElementNode` method, but `getPreviousSibling()` can return a block-level `DecoratorNode`, which lacks it → TypeError. The anchor-side branch has the same latent flaw: it doesn't throw, but sets an invalid element point on a decorator (which then breaks downstream selection traversal).

## Fix

Both branches now resolve the shrink target via `$nearestElementSibling(block, direction)`, which walks the sibling chain past non-element nodes (decorators) to the nearest `ElementNode`. A flush endpoint therefore lands beyond the decorator, so an edge paragraph with no selected content is not formatted — preserving #1153's behavior across decorator boundaries (the decorator acts as a transparent boundary, matching what the same selection produces between adjacent paragraphs). If no element sibling exists in that direction, the endpoint stays put, which safely degrades to the pre-#1153 wrap for that edge only.

(First iteration guarded with plain `$isElementNode` and skipped the shrink next to decorators; review feedback pointed out that this dropped #1153's improvement for decorator-adjacent selections, so the guard became a walk.)

## Origin

Introduced by c2bf790c "Don't absorb the block above a code block / quote selection" (#1153), 2026-06-23, first shipped in v0.9.20.

## Impact (from Basecamp, the consuming app)

Sentry [BC3-JS-N8J4](https://basecamp.sentry.io/issues/7569588601/): 809 events / 563 users since Jun 23, steady ~40/day, plus fingerprint-split sibling BC3-JS-N8QA.

## Testing

- Added `test/browser/tests/formatting/block_format_beside_decorator.test.js` (Playwright E2E; moved from a jsdom unit test per review): heading and quote formats applied through the real toolbar via the editor handle, with multi-block selections flush against a horizontal divider, plus a both-blocks-selected case. The flush-endpoint scenarios reproduce the production TypeError against the unguarded code, and pin the walk-past semantics (edge block with no selected content stays unformatted; `<hr>` survives).
- `editor_handle.js`'s `clickToolbarButton` now opens the "format" dropdown for heading/Paragraph/Clear-formatting commands, so those formats can be driven through the handle.
- Browser suites, vitest unit suite, and lint green in CI.
- Manually validated against the production build vs. this branch (`manually-validate-fix`) — labeled `fix-validated`.

## Deploy path

Needs a lexxy release + version bump in bc3 to reach production.

Tracking card: https://app.basecamp.com/2914079/buckets/27/card_tables/cards/10092012358

### CODE REVIEW

File: src/helpers/lexical_helper.js [hunk 1]
```diff
@@ -274,12 +274,12 @@ function $shrinkSelectionPastBlockEdges(selection) {
 
   if ($isAtBlockEnd(selection.anchor, anchorBlock)) {
     const nextBlock = anchorBlock.getNextSibling()
-    if (nextBlock) selection.anchor.set(nextBlock.getKey(), 0, "element")
+    if ($isElementNode(nextBlock)) selection.anchor.set(nextBlock.getKey(), 0, "element")
```
[the remark this answers is not in the document]
Comment (jeremy) [reply src/helpers/lexical_helper.js]: 🤖 Fixed in 3af491fd — the shrink now walks past non-element siblings to the nearest ElementNode instead of abandoning the move, so flush endpoints land beyond the decorator and edge paragraphs with no selected content are no longer formatted. In `<p>alpha</p><hr><p>beta</p>`, selecting from end-of-alpha through beta now formats only beta — same behavior as plain adjacent paragraphs. Tests updated to assert the edge block stays unformatted (verified they fail against the plain guard), plus a both-blocks-selected case.

**Comment by jeremy** [comment]
@codex review

File: test/javascript/unit/editor/block_format_beside_decorator.test.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+function selectFromFirstTextToLastText(editorElement, { anchorAtEnd = false, focusAtEnd = false } = {}) {
+  editorElement.editor.update(() => {
+    const root = $getRoot()
+    const firstText = root.getFirstChild().getFirstDescendant()
+    const lastText = root.getLastChild().getFirstDescendant()
+    const anchorOffset = anchorAtEnd ? firstText.getTextContentSize() : 0
+    const focusOffset = focusAtEnd ? lastText.getTextContentSize() : 0
+    const selection = firstText.select(anchorOffset, anchorOffset)
+    selection.focus.set(lastText.getKey(), focusOffset, "text")
```
[the remark this answers is not in the document]
Comment (jeremy) [reply test/javascript/unit/editor/block_format_beside_decorator.test.js]: 🤖 Fixed in 7cbfe159 — the helper now uses root.getFirstDescendant()/getLastDescendant() with a null guard, mirroring selectAll() in editor_helper.js.

File: src/helpers/lexical_helper.js [hunk 3]
```diff
@@ -273,16 +273,27 @@ function $shrinkSelectionPastBlockEdges(selection) {
...
-    const nextBlock = anchorBlock.getNextSibling()
+    const nextBlock = $nearestElementSibling(anchorBlock, "next")
     if (nextBlock) selection.anchor.set(nextBlock.getKey(), 0, "element")
   }
 
   if ($isAtBlockStart(selection.focus, focusBlock)) {
-    const previousBlock = focusBlock.getPreviousSibling()
+    const previousBlock = $nearestElementSibling(focusBlock, "previous")
     if (previousBlock) selection.focus.set(previousBlock.getKey(), previousBlock.getChildrenSize(), "element")
   }
```
[the remark this answers is not in the document]
Comment (jeremy) [reply src/helpers/lexical_helper.js:283]: 🤖 Fixed — the PR description now describes the walk-past behavior (with a note on the guard→walk evolution from review feedback); the code keeps the walk-past semantics.

**Comment by jeremy** [comment]
@codex review

File: src/helpers/lexical_helper.js [hunk 4]
```diff
@@ -273,16 +273,27 @@ function $shrinkSelectionPastBlockEdges(selection) {
   if (!anchorBlock || !focusBlock || anchorBlock.is(focusBlock)) return
 
   if ($isAtBlockEnd(selection.anchor, anchorBlock)) {
-    const nextBlock = anchorBlock.getNextSibling()
+    const nextBlock = $nearestElementSibling(anchorBlock, "next")
     if (nextBlock) selection.anchor.set(nextBlock.getKey(), 0, "element")
   }
```
[the remark this answers is not in the document]
Comment (jeremy) [reply src/helpers/lexical_helper.js:278]: 🤖 Fixed in 1fca13a4 — $shrinkSelectionPastBlockEdges now calls $ensureForwardRangeSelection at the end, so crossed endpoints from a both-edges-flush shrink are re-normalized before callers treat anchor/focus as start/end.

File: test/javascript/unit/editor/block_format_beside_decorator.test.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    expect(editorElement.value).toContain("<h2>alpha</h2>")
+    expect(editorElement.value).toContain("<h2>beta</h2>")
+    expect(editorElement.value).toContain("<hr>")
+
+    await destroyTestEditor(editorElement)
+  })
+
+  // Both endpoints flush against the divider selects only the divider itself.
+  // This behaves the same as the equivalent selection across two adjacent
+  // paragraphs with no divider between them: everything gets wrapped.
```
[the remark this answers is not in the document]
Comment (jeremy) [reply test/javascript/unit/editor/block_format_beside_decorator.test.js]: 🤖 Fixed in 1fca13a4 — the test now pins the exact output (<blockquote><p>alpha</p><hr><p>beta</p></blockquote>), proving the wrap the comment describes.

**Comment by jeremy** [comment]
@codex review

File: test/javascript/unit/editor/block_format_beside_decorator.test.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
+import { describe, expect, test } from "vitest"
```
Comment (jorgemanrubia) [inline test/javascript/unit/editor/block_format_beside_decorator.test.js]: These test should go in `test/browser/tests/formatting` (end-to-end tests).

File: test/javascript/unit/editor/block_format_beside_decorator.test.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import { $getRoot } from "lexical"
+import { createTestEditor, destroyTestEditor, dispatchToolbarCommand, setContent, tick } from "../helpers/editor_helper"
+
+function selectFromFirstTextToLastText(editorElement, { anchorAtEnd = false, focusAtEnd = false } = {}) {
```
Comment (jorgemanrubia) [inline test/javascript/unit/editor/block_format_beside_decorator.test.js]: Maybe this could be placed in `/home/jorge/Work/basecamp/lexxy/test/browser/helpers/editor_handle.js` and use the editor handle here?

File: test/javascript/unit/editor/block_format_beside_decorator.test.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  }, { discrete: true })
+}
+
+describe("block formats next to decorator nodes", () => {
+  test("focus flush after a horizontal divider walks back past it, leaving the trailing block unformatted", async () => {
+    const editorElement = await createTestEditor()
+    await setContent(editorElement, "<p>alpha</p><hr><p>beta</p>")
+    selectFromFirstTextToLastText(editorElement)
+
+    expect(() => dispatchToolbarCommand(editorElement, "applyHeadingFormat", "h2")).not.toThrow()
```
Comment (jorgemanrubia) [inline test/javascript/unit/editor/block_format_beside_decorator.test.js]: I know this dispatchToolbarCommadn has sneaked in from other tests, but if using the editor handle for tests we can use `editor.clickToolbarButton` instead.

**Review by jorgemanrubia — COMMENTED** [review]
👏

**Comment by jorgemanrubia** [comment]
## Manual Validation: Validated

🤖 Ran `/manually-validate-fix` — one Selenium script, driven like a human against both environments with identical steps and selection geometry.

**Reproduction:** built `<p>alpha</p><hr><p>beta</p>` in the editor (typed `alpha`, Enter, `---`, Enter, `beta`), then mouse-dragged a selection from inside `alpha` down to the far left of `beta`. Live selection: anchor `(alpha, 1)`, focus `(beta, 0)` — focus flush at the start of a block whose previous top-level sibling is the `<hr>` divider decorator. Then clicked the Quote toolbar button.

**Production** (`https://basecamp.github.io/lexxy/`, `@37signals/lexxy@latest` via esm.sh):
- Quote silently did nothing — editor value unchanged (`<p>alpha</p><hr><p>beta</p>`), no `<blockquote>`, button not active.
- Console: `TypeError: i.getChildrenSize is not a function` — the exact production error.

**Local** (`http://lexxy.localhost:3000`, this branch `bugs/BC-10092012358`):
- Quote applied cleanly → `<blockquote><p>alpha</p></blockquote><hr /><p>beta</p>`, divider and `beta` intact.
- No `getChildrenSize` / TypeError in console.

The `$nearestElementSibling` walk past the decorator sibling resolves the throw and lets the block format apply. Verdict: broken in production, fixed on this branch.

File: test/javascript/unit/editor/block_format_beside_decorator.test.js (hunk 6 above)
Comment (jorgemanrubia) [reply test/javascript/unit/editor/block_format_beside_decorator.test.js]: 🤖 Moved the decorator regression tests out of `test/javascript/unit/` into the browser E2E suite at `test/browser/tests/formatting/block_format_beside_decorator.test.js` (commit b5f6d27b). They now drive the real toolbar in a browser instead of dispatching Lexical commands in jsdom.

File: test/javascript/unit/editor/block_format_beside_decorator.test.js (hunk 7 above)
Comment (jorgemanrubia) [reply test/javascript/unit/editor/block_format_beside_decorator.test.js]: 🤖 Done — the tests now use the `editor` handle (from `test/browser/helpers/editor_handle.js`). Selections are built through the handle and formats are applied via `editor.clickToolbarButton(...)`.

File: test/javascript/unit/editor/block_format_beside_decorator.test.js (hunk 8 above)
Comment (jorgemanrubia) [reply test/javascript/unit/editor/block_format_beside_decorator.test.js]: 🤖 Fixed — `dispatchToolbarCommand` is gone entirely. It was removed from the shared unit helper in the separate refactor PR (#1198, now merged), and these tests use `editor.clickToolbarButton` instead. The editor handle's `clickToolbarButton` now also opens the "format" dropdown for heading / Paragraph / Clear-formatting commands, since those buttons live inside the hidden dropdown panel.

**Comment by jeremy** [comment]
@codex review

File: test/browser/tests/formatting/block_format_beside_decorator.test.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
+import { test } from "../../test_helper.js"
+import { assertEditorHtml } from "../../helpers/assertions.js"
+
+// Selects from the first text node to the last text node in the editor,
+// mirroring the production selection shapes that border a decorator: an
+// endpoint sitting flush against a block edge (offset 0 or the end of the
+// text) lands next to the <hr>, which is the case that used to throw.
+async function selectAcrossDivider(editor, { anchorAtEnd = false, focusAtEnd = false } = {}) {
```
Comment (jorgemanrubia) [inline test/browser/tests/formatting/block_format_beside_decorator.test.js]: Please move function to the end of the test file

File: test/browser/tests/formatting/block_format_beside_decorator.test.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
+import { test } from "../../test_helper.js"
+import { assertEditorHtml } from "../../helpers/assertions.js"
+
```
[the remark this answers is not in the document]
Comment (jeremy) [reply test/browser/tests/formatting/block_format_beside_decorator.test.js:3]: 🤖 Fixed — the Testing section now describes the Playwright browser test at `test/browser/tests/formatting/block_format_beside_decorator.test.js` and the editor-handle change; the stale unit-test reference is gone.

File: test/browser/tests/formatting/block_format_beside_decorator.test.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+  test("selecting content in both blocks around a horizontal divider formats both and keeps the divider", async ({ editor }) => {
+    await editor.setValue("<p>alpha</p><hr><p>beta</p>")
+    await selectAcrossDivider(editor, { focusAtEnd: true })
+
+    await editor.clickToolbarButton("heading-large")
+
+    await assertEditorHtml(editor, "<h2>alpha</h2><hr><h2>beta</h2>")
+  })
+
+  // Both endpoints flush against the divider selects only the divider itself.
```
[the remark this answers is not in the document]
Comment (jeremy) [reply test/browser/tests/formatting/block_format_beside_decorator.test.js]: 🤖 Fixed in 5d417991 — comment reworded.

**Comment by jeremy** [comment]
@codex review

File: test/browser/tests/formatting/block_format_beside_decorator.test.js (hunk 9 above)
Comment (jorgemanrubia) [reply test/browser/tests/formatting/block_format_beside_decorator.test.js]: 🤖 Done — moved `selectAcrossDivider` below the `test.describe` block, at the end of the file (commit a04da849). Function hoisting keeps it available to the tests. Suite still green on chromium + firefox.

File: test/browser/tests/formatting/block_format_beside_decorator.test.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+  await editor.content.evaluate((el, { anchorAtEnd, focusAtEnd }) => {
+    const walker = document.createTreeWalker(el, NodeFilter.SHOW_TEXT)
+    const textNodes = []
+    let node
+    while ((node = walker.nextNode())) textNodes.push(node)
+    const firstText = textNodes[0]
+    const lastText = textNodes[textNodes.length - 1]
+
+    const range = document.createRange()
+    range.setStart(firstText, anchorAtEnd ? firstText.nodeValue.length : 0)
```
[the remark this answers is not in the document]
Comment (jorgemanrubia) [reply test/browser/tests/formatting/block_format_beside_decorator.test.js]: 🤖 Good call — added an explicit guard in `selectAcrossDivider` (commit bf10bb97): it now throws a clear "needs a fixture with at least one text node" error instead of a cryptic DOM Range failure if the helper is ever reused with a text-node-less document.

File: test/browser/tests/formatting/block_format_beside_decorator.test.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+test.describe("Block formats next to decorator nodes", () => {
+  test.beforeEach(async ({ page }) => {
+    await page.goto("/")
+    await page.waitForSelector("lexxy-editor[connected]")
+    await page.waitForSelector("lexxy-toolbar[connected]")
+  })
+
+  test("focus flush after a horizontal divider walks back past it, leaving the trailing block unformatted", async ({ editor }) => {
+    await editor.setValue("<p>alpha</p><hr><p>beta</p>")
+    await selectAcrossDivider(editor)
```
[the remark this answers is not in the document]
Comment (jorgemanrubia) [reply test/browser/tests/formatting/block_format_beside_decorator.test.js:13]: 🤖 I looked into adding an attachment-boundary case and decided against it, for a concrete reason. The fix is decorator-type-agnostic: `$shrinkSelectionPastBlockEdges` walks top-level *siblings* via `$nearestElementSibling`, which stops at any non-`ElementNode` — `HorizontalDividerNode`, `ActionTextAttachmentNode`, and `CustomActionTextAttachmentNode` are all just "not an ElementNode" to that walk, so `<hr>` and an attachment exercise the identical code path.

More importantly, I built the attachment test (top-level `<action-text-attachment>` between two paragraphs on `/attachments-enabled.html`, same flush selection) and verified it against the *pre-fix* code — it did **not** reproduce the `getChildrenSize` throw. Because the attachment renders as a `contenteditable="false"` `<figure>`, the browser normalizes a DOM range flush against it differently than against an `<hr>`, so the resulting Lexical selection never hits the crashing flush branch. The test therefore passed on both the broken and fixed code — a false-confidence test. I would rather not add a regression test that does not actually guard the regression. The `<hr>` cases (verified to fail against the unguarded code) do cover the shared walk-past logic.

[the remark this answers is not in the document]
Comment (jeremy) [reply test/browser/tests/formatting/block_format_beside_decorator.test.js:13]: 🤖 Confirmed the false-confidence result independently, and found the mechanism: a DOM range spanning the attachment's `contenteditable="false"` figure normalizes to a **NodeSelection**, which block-format commands no-op on (`$isRangeSelection` guard in `Contents`) — so a DOM-built selection never reaches the flush branch on broken *or* fixed code.

The flush shape is constructible by building the RangeSelection in Lexical instead (same pattern as the handle's `placeCaretOnQuoteElement`): anchor `alpha:0`, focus `beta:0` inside `editor.update`. That version does guard the regression — verified it fails against the pre-fix helper (heading never applies) and passes with the fix. I have the patch parked locally (a `selectFromTextToText` handle method + one test) — happy to push it if you want the attachment flavor pinned; equally fine leaving the `<hr>` cases to cover the shared walk, per your call.

[the remark this answers is not in the document]
Comment (jorgemanrubia) [reply test/browser/tests/formatting/block_format_beside_decorator.test.js:13]: 🤖 Yes please — push it. Pinning the attachment flavor with the programmatic RangeSelection (the `selectFromTextToText` handle method) is worth having as a real regression guard, since it genuinely fails pre-fix and passes with the fix. Thanks for digging into the NodeSelection mechanism. The `<hr>` cases can stay as-is alongside it.
